### PR TITLE
Ava 148

### PR DESCRIPTION
**Pull request by aterentic-ethernal**
This is partial app client implementation, it only app specific cells from IPFS and RPC. Since there are a lot of cleanup that needs to be done, doing feature partially should enable parallel development of other features.

### CODE REVIEW

File: src/app_client.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	rpc_url: String,
+	app_id: u32,
+	block_receive: Receiver<ClientMsg>,
+) {
+	log::info!("Starting for app {}...", app_id);
+
+	for block in block_receive {
+		log::info!("Block {} available", block.number);
+		if let Err(error) = process_block(&ipfs, &rpc_url, app_id, &block).await {
+			log::error!("Cannot process block {}: {}", block.number, error);
```
Comment (sh3ll3x3c) [inline src/app_client.rs:66]: Perhaps we could add an explicit `continue` here to make sure we don't overlook the error edge case once we add further functionalities (ie. cell IPFS upload after reconstruction)

File: src/util.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	let (app_ids, offsets): (Vec<_>, Vec<_>) = index.iter().cloned().unzip();
+	// Prepend app_id zero
+	let mut app_ids_ext = vec![0];
+	app_ids_ext.extend(app_ids);
+
+	// Prepend offset 0 for app_id 0
+	let mut offsets_ext = vec![0];
+	offsets_ext.extend(offsets);
+
+	let mut sizes = offsets_ext[0..offsets_ext.len() - 1]
```
Comment (sh3ll3x3c) [inline src/util.rs:15]: Well thought off :)

File: src/data.rs [hunk 3]
```diff
@@ -453,6 +468,73 @@ pub fn decode_block_cid_ask_message(msg: Vec<u8>) -> Option<(i128, Option<Cid>)>
...
+			block: block_number,
+			row: position.row,
+			col: position.col,
+			proof: block.data().to_vec(),
+			data: block.data()[48..].to_vec(),
+		})
+	})
+}
+
+pub async fn fetch_cells_from_ipfs(
```
Comment (sh3ll3x3c) [inline src/data.rs:498]: We should change (or remove) the `ipfs_priority_get_cells` function in the LC, to make sure we have uniform approach to data retrieval.

File: src/util.rs (hunk 2 above)
Comment (aterentic-ethernal) [reply src/util.rs:15]: 👍
We have to replace layouts with offsets and remove this for good.

File: src/data.rs [hunk 4]
```diff
@@ -453,6 +468,73 @@ pub fn decode_block_cid_ask_message(msg: Vec<u8>) -> Option<(i128, Option<Cid>)>
...
+	log::debug!("Number of known IPFS peers: {}", peers.len());
+
+	if peers.is_empty() {
+		log::info!("No known IPFS peers");
+		return Ok((vec![], positions.to_vec()));
+	}
+
+	let mut fetched = vec![];
+	let mut unfetched = vec![];
+	// TODO: Paralelize fetching
```
[the remark this answers is not in the document]
Comment (aterentic-ethernal) [reply src/data.rs:511]: I guess you mean once we implement concurrent fetch?

File: src/app_client.rs (hunk 1 above)
Comment (aterentic-ethernal) [reply src/app_client.rs:74]: Done.

**Pull request merged**